doc/design.md

```diff
@@ -0,0 +1,1268 @@
+  See the [C++ example](#examples-cpp) for an example of how this would be used.
+
+The current implementation seems to support all required features.
+Performance was originally a problem due to the multi-dimensional iteration, but has subsequently been optimized to the point where it should be sufficient for most applications.
+
+
+### <a name="design-components-datasetindex"></a>`class DatasetIndex`
+
+`class DatasetIndex` is a simple helper to index slices in a dataset via values in an axis (coordinate).
+`Dataset` itself makes no attempt at keeping values in a coordinate variable validated.
+For example, there is no check for duplicate spectrum numbers.
+Therefore, direct indexing is not possible.
+Instead `DatasetIndex` builds the index (and does the validation) on demand when constructed.
+
+##### Rationale
+
+I tried to *maintain* consistent indices like spectrum numbers using code in `Indexing::IndexInfo`, but it turned out to be really complicated when integrating it into `API::MatrixWorkspace`.
+
+
+### <a name="design-components-units"></a>Units
+
+Internally the prototype uses `boost::units`, but we require run-time units.
+It is therefore wrapped in a class that simply stores an unique ID for every unit.
+- Arithmetic operations branch off using the ID, relying on `boost::units` for the actual work.
+- Probably need to add support for units scales.
+
+Very little of this is implemented in the prototype apart from the most basic examples.
+This will probably require a not insignificant amount of work to get right to make sure we capture all the subtleties.
+See for example [An issue with distributions and dimensionless units](https://github.com/mantidproject/documents/blob/fix-divide-distribution/Design/DistributionsAndDimensionlessData.md) for a recent issue relating to this in Mantid.
+
+
+### <a name="design-components-exceptions"></a>Exceptions
+
+The prototype currently simply throws `std::runtime_error` in most cases.
+We should most likely add a clear range of exception types that provide useful information such as indices about the failure.
+
+
+## <a name="examples"></a>Examples
+
+
+### <a name="examples-cpp"></a>C++ example
+
+The following is a relatively complex example of the capabilities of `Dataset`.
+Apart from minor details and operations that have not been fully implemented yet, all features that are used below have been demonstrated to work in the prototype.
+
+```cpp
+// Create an empty dataset.
+Dataset d;
+
+// Add some information about the instrument. This would likely be wrapped in
+// convenience functions.
+// Data passed as argument, e.g., initializer_list.
+d.insert<Coord::DetectorId>({Dim::Detector, 4}, {1, 2, 3, 4});
+// No data passed, will be default-constructed.
+d.insert<Coord::DetectorPosition>({Dim::Detector, 4});
+
+// Add spectrum-to-detector mapping and spectrum numbers.
+std::vector<std::vector<gsl::index>> grouping = {{0, 2}, {1}, {}};
```

Comment:
Is `gsl::index` from the ``guidelines support library? Is the suggestion to include that project?

Reply:
Yes. Yes, potentially, or just provide our own. I am also using `gsl::span` (`std::span` is only C++20).